src/webapi/gamepad.rs

```diff
@@ -0,0 +1,217 @@
+    }
+
+    /// An auto-incrementing integer to uniquely identify a connected Gamepad.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/index)
+    // https://w3c.github.io/gamepad/#dom-gamepad-index
+    #[inline]
+    fn index(&self) -> i32 {
+        js!(
+            return @{self.as_ref()}.index;
+        ).try_into().unwrap()
+    }
+
+    /// Is this gamepad connected to the system, powered on, and usable?
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/connected)
+    // https://w3c.github.io/gamepad/#dom-gamepad-connected
+    #[inline]
+    fn connected(&self) -> bool {
+        js!(
+            return @{self.as_ref()}.connected;
+        ).try_into().unwrap()
+    }
+
+    /// Monotonically increasing time this gamepad was updated.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/timestamp)
+    // https://w3c.github.io/gamepad/#dom-gamepad-timestamp
+    #[inline]
+    fn timestamp(&self) -> f64 {
+        js!(
+            return @{self.as_ref()}.timestamp;
+        ).try_into().unwrap()
+    }
+
+    /// The mapping in use for this device.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/mapping)
+    // https://w3c.github.io/gamepad/#dom-gamepad-mapping
+    #[inline]
+    fn mapping(&self) -> GamepadMappingType {
+        js!(
+            return @{self.as_ref()}.mapping;
+        ).try_into().unwrap()
+    }
+
+    /// Array of values for all axes of the gamepad.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/axes)
+    // https://w3c.github.io/gamepad/#dom-gamepad-axes
+    #[inline]
+    fn axes(&self) -> Vec<f64> {
+        js!(
+            return @{self.as_ref()}.axes;
+        ).try_into().unwrap()
+    }
+
+    /// Array of button states for all buttons of the gamepad.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/buttons)
+    // https://w3c.github.io/gamepad/#dom-gamepad-buttons
+    #[inline]
+    fn buttons(&self) -> Vec<GamepadButton> {
+        js!(
+            return @{self.as_ref()}.buttons;
+        ).try_into().unwrap()
+    }
+}
+
+/// A reference to a JavaScript object which implements the [IGamepad](trait.IGamepad.html)
+/// interface.
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad)
+// https://w3c.github.io/gamepad/#gamepad-interface
+#[derive(Clone, Debug, Eq, PartialEq, ReferenceType)]
+#[reference(instance_of = "Gamepad")]
+pub struct Gamepad( Reference );
+
+impl IGamepad for Gamepad {}
+
+/// Retrieve all connected gamepads, in an array indexed by each gamepad's `index` member.
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/getGamepads)
+// https://w3c.github.io/gamepad/#dom-gamepad-axes
+pub fn get_gamepads() -> Vec<Option<Gamepad>> {
```

Comment:
This would be better as `Gamepad::get_all` or something like that.

Reply:
If we absolutely want to adhere to the specs we could add a `navigator` method (just as we do for e.g. [document](https://docs.rs/stdweb/*/stdweb/web/fn.document.html)) and then define `get_gamepads` on it. Then it would be obvious that it's a 1-to-1 call.

In extra convenience methods like this one we usually just pick a name that's more Rust-y, and in this case since the call will be prefixed with `Gamepad::` there is no point in repeating it. 

That said, it's not that much of a big deal, so I leave this decision to you.

Reply:
Okay. `Gamepad::get_all()` does sound nicer. I renamed to `Gamepad::get_all()`, and documented it should be called every frame.